File: desktop/lib/models/product.dart
import 'entities.dart';

class ProductAttributeValueRecord {
  const ProductAttributeValueRecord({
    required this.id,
    required this.attributeDefinitionId,
    required this.valueText,
    required this.valueNumber,
    required this.valueDate,
    required this.valueBoolean,
  });

  final String id;
  final String attributeDefinitionId;
  final String valueText;
  final String valueNumber;
  final String valueDate;
  final bool? valueBoolean;

  factory ProductAttributeValueRecord.fromJson(Json json) =>
      ProductAttributeValueRecord(
        id: stringValue(json['id']),
        attributeDefinitionId: stringValue(json['attribute_definition_id']),
        valueText: stringValue(json['value_text']),
        valueNumber: stringValue(json['value_number']),
        valueDate: stringValue(json['value_date']),
        valueBoolean: json['value_boolean'] is bool
            ? json['value_boolean'] as bool
            : null,
      );
}

class ProductMediaRecord {
  const ProductMediaRecord({
    required this.id,
    required this.mediaKind,
    required this.fileName,
    required this.mimeType,
    required this.storagePath,
    required this.isPrimary,
  });

  final String id;
  final String mediaKind;
  final String fileName;
  final String mimeType;
  final String storagePath;
  final bool isPrimary;

  factory ProductMediaRecord.fromJson(Json json) => ProductMediaRecord(
        id: stringValue(json['id']),
        mediaKind: stringValue(json['media_kind']),
        fileName: stringValue(json['file_name']),
        mimeType: stringValue(json['mime_type']),
        storagePath: stringValue(json['storage_path']),
        isPrimary: boolValue(json['is_primary']),
      );
}

class ProductCategoryRecord {
  const ProductCategoryRecord({
    required this.id,
    required this.code,
    required this.name,
    required this.parentId,
    required this.level,
    required this.path,
    required this.isActive,
  });

  final String id;
  final String code;
  final String name;
  final String parentId;
  final int level;
  final String path;
  final bool isActive;

  factory ProductCategoryRecord.fromJson(Json json) => ProductCategoryRecord(
        id: stringValue(json['id']),
        code: stringValue(json['code']),
        name: stringValue(json['name']),
        parentId: stringValue(json['parent_id']),
        level: (json['level'] as num?)?.toInt() ?? 0,
        path: stringValue(json['path']),
        isActive: boolValue(json['is_active'], fallback: true),
      );
}

class ProductFeatureState {
  const ProductFeatureState({required this.code, required this.enabled});

  final String code;
  final bool enabled;

  factory ProductFeatureState.fromJson(Json json) => ProductFeatureState(
        code: stringValue(json['code']),
        enabled: boolValue(json['enabled']),
      );
}

class ProductTaxProfileRecord {
  const ProductTaxProfileRecord({
    required this.id,
    required this.code,
    required this.groupCode,
    required this.label,
    required this.taxSystemId,
  });

  final String id;
  final String code;
  final String groupCode;
  final String label;
  final String taxSystemId;

  factory ProductTaxProfileRecord.fromJson(Json json) =>
      ProductTaxProfileRecord(
        id: stringValue(json['id']),
        code: stringValue(json['code']),
        groupCode: stringValue(json['group_code']),
        label: stringValue(json['label']),
        taxSystemId: stringValue(json['tax_system_id']),
      );
}

class ProductMetadataRecord {
  const ProductMetadataRecord({
    required this.profileCode,
    required this.features,
    required this.categories,
    required this.taxProfiles,
    required this.requiredAttributeDefinitionIds,
    required this.optionalAttributeDefinitionIds,
  });

  final String profileCode;
  final List<ProductFeatureState> features;
  final List<ProductCategoryRecord> categories;
  final List<ProductTaxProfileRecord> taxProfiles;
  final List<String> requiredAttributeDefinitionIds;
  final List<String> optionalAttributeDefinitionIds;

  bool featureEnabled(String code) =>
      features.any((item) => item.code.toUpperCase() == code && item.enabled);

  factory ProductMetadataRecord.fromJson(Json json) => ProductMetadataRecord(
        profileCode: stringValue(json['profile_code']),
        features: _objects(json['features'])
            .map(ProductFeatureState.fromJson)
            .toList(),
        categories: _objects(json['categories'])
            .map(ProductCategoryRecord.fromJson)
            .toList(),
        taxProfiles: _objects(json['tax_profiles'])
            .map(ProductTaxProfileRecord.fromJson)
            .toList(),
        requiredAttributeDefinitionIds:
            stringList(json['required_attribute_definition_ids']),
        optionalAttributeDefinitionIds:
            stringList(json['optional_attribute_definition_ids']),
      );
}

class Product {
  const Product({
    required this.id,
    required this.firmId,
    required this.code,
    required this.barcode,
    required this.qrCode,
    required this.name,
    required this.shortName,
    required this.description,
    required this.productType,
    required this.categoryId,
    required this.subCategoryId,
    required this.unit,
    required this.brand,
    required this.model,
    required this.hsnSac,
    this.taxProfileGroupCode = '',
    String? taxProfileId,
    this.baseUomId = '',
    this.inventoryUomId = '',
    this.purchaseUomId = '',
    this.salesUomId = '',
    this.defaultReceivingUomId = '',
    this.defaultDispatchUomId = '',
    this.minimumSalesUomId = '',
    this.weight = '',
    this.volume = '',
    this.length = '',
    this.width = '',
    this.height = '',
    this.allowFraction = false,
    this.allowDecimal = true,
    required this.purchasePrice,
    required this.sellingPrice,
    required this.mrp,
    required this.status,
    required this.remarks,
    this.trackBatch = false,
    this.trackLot = false,
    this.trackSerial = false,
    this.trackExpiry = false,
    this.trackManufacturingDate = false,
    this.trackWarranty = false,
    this.allowNegativeStock = false,
    this.requireBatchOnReceipt = false,
    this.requireBatchOnIssue = false,
    this.requireSerialOnReceipt = false,
    this.requireSerialOnIssue = false,
    required this.isDeleted,
    required this.createdAt,
    required this.updatedAt,
    required this.attributes,
    required this.media,
  });

  final String id;
  final String firmId;
  final String code;
  final String barcode;
  final String qrCode;
  final String name;
  final String shortName;
  final String description;
  final String productType;
  final String categoryId;
  final String subCategoryId;
  final String unit;
  final String brand;
  final String model;
  final String hsnSac;
  final String taxProfileGroupCode;
  final String baseUomId;
  final String inventoryUomId;
  final String purchaseUomId;
  final String salesUomId;
  final String defaultReceivingUomId;
  final String defaultDispatchUomId;
  final String minimumSalesUomId;
  final String weight;
  final String volume;
  final String length;
  final String width;
  final String height;
  final bool allowFraction;
  final bool allowDecimal;
  final String purchasePrice;
  final String sellingPrice;
  final String mrp;
  final String status;
  final String remarks;
  final bool trackBatch;
  final bool trackLot;
  final bool trackSerial;
  final bool trackExpiry;
  final bool trackManufacturingDate;
  final bool trackWarranty;
  final bool allowNegativeStock;
  final bool requireBatchOnReceipt;
  final bool requireBatchOnIssue;
  final bool requireSerialOnReceipt;
  final bool requireSerialOnIssue;
  final bool isDeleted;
  final String createdAt;
  final String updatedAt;
  final List<ProductAttributeValueRecord> attributes;
  final List<ProductMediaRecord> media;

  factory Product.fromJson(Json json) => Product(
        id: stringValue(json['id']),
        firmId: stringValue(json['firm_id']),
        code: stringValue(json['code']),
        barcode: stringValue(json['barcode']),
        qrCode: stringValue(json['qr_code']),
        name: stringValue(json['name']),
        shortName: stringValue(json['short_name']),
        description: stringValue(json['description']),
        productType: stringValue(json['product_type']),
        categoryId: stringValue(json['category_id']),
        subCategoryId: stringValue(json['sub_category_id']),
        unit: stringValue(json['unit']),
        brand: stringValue(json['brand']),
        model: stringValue(json['model']),
        hsnSac: stringValue(json['hsn_sac']),
        taxProfileGroupCode:
            stringValue(json['tax_profile_group_code']).isNotEmpty
                ? stringValue(json['tax_profile_group_code'])
                : stringValue(json['tax_profile_id']),
        baseUomId: stringValue(json['base_uom_id']),
        inventoryUomId: stringValue(json['inventory_uom_id']),
        purchaseUomId: stringValue(json['purchase_uom_id']),
        salesUomId: stringValue(json['sales_uom_id']),
        defaultReceivingUomId: stringValue(json['default_receiving_uom_id']),
        defaultDispatchUomId: stringValue(json['default_dispatch_uom_id']),
        minimumSalesUomId: stringValue(json['minimum_sales_uom_id']),
        weight: stringValue(json['weight']),
        volume: stringValue(json['volume']),
        length: stringValue(json['length']),
        width: stringValue(json['width']),
        height: stringValue(json['height']),
        allowFraction: boolValue(json['allow_fraction']),
        allowDecimal: boolValue(json['allow_decimal'], fallback: true),
        purchasePrice: stringValue(json['purchase_price']),
        sellingPrice: stringValue(json['selling_price']),
        mrp: stringValue(json['mrp']),
        status: stringValue(json['status']),
        remarks: stringValue(json['remarks']),
        trackBatch: boolValue(json['track_batch']),
        trackLot: boolValue(json['track_lot']),
        trackSerial: boolValue(json['track_serial']),
        trackExpiry: boolValue(json['track_expiry']),
        trackManufacturingDate: boolValue(json['track_manufacturing_date']),
        trackWarranty: boolValue(json['track_warranty']),
        allowNegativeStock: boolValue(json['allow_negative_stock']),
        requireBatchOnReceipt: boolValue(json['require_batch_on_receipt']),
        requireBatchOnIssue: boolValue(json['require_batch_on_issue']),
        requireSerialOnReceipt: boolValue(json['require_serial_on_receipt']),
        requireSerialOnIssue: boolValue(json['require_serial_on_issue']),
        isDeleted: boolValue(json['is_deleted']),
        createdAt: stringValue(json['created_at']),
        updatedAt: stringValue(json['updated_at']),
        attributes: _objects(json['attributes'])
            .map(ProductAttributeValueRecord.fromJson)
            .toList(),
        media:
            _objects(json['media']).map(ProductMediaRecord.fromJson).toList(),
      );
}

class ProductQuery {
  const ProductQuery({
    this.status,
    this.productType,
    this.categoryId,
    this.taxProfileGroupCode,
    this.brand,
    this.hsnSac,
    this.attributeQuery,
    this.includeDeleted = false,
  });

  final String? status;
  final String? productType;
  final String? categoryId;
  final String? taxProfileGroupCode;
  final String? brand;
  final String? hsnSac;
  final String? attributeQuery;
  final bool includeDeleted;

  Map<String, String> toQuery() => {
        if (status?.isNotEmpty == true) 'status': status!,
        if (productType?.isNotEmpty == true) 'product_type': productType!,
        if (categoryId?.isNotEmpty == true) 'category_id': categoryId!,
        if (taxProfileGroupCode?.isNotEmpty == true)
          'tax_profile_group_code': taxProfileGroupCode!,
        if (brand?.isNotEmpty == true) 'brand': brand!,
        if (hsnSac?.isNotEmpty == true) 'hsn_sac': hsnSac!,
        if (attributeQuery?.isNotEmpty == true)
          'attribute_query': attributeQuery!,
        if (includeDeleted) 'include_deleted': 'true',
      };
}

List<Json> _objects(dynamic value) => value is List
    ? value
        .whereType<Map>()
        .map((item) => Map<String, dynamic>.from(item))
        .toList()
    : const [];

File: backend/app/products/schemas/product.py
"""Validated request and response contracts for product management."""

from datetime import date, datetime
from decimal import Decimal
from enum import StrEnum
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator


class ProductType(StrEnum):
    """Supported product type classifications."""

    STOCK_ITEM = "STOCK_ITEM"
    SERVICE = "SERVICE"
    RAW_MATERIAL = "RAW_MATERIAL"
    FINISHED_GOODS = "FINISHED_GOODS"
    SEMI_FINISHED = "SEMI_FINISHED"
    ASSET = "ASSET"
    CONSUMABLE = "CONSUMABLE"
    BUNDLE = "BUNDLE"
    DIGITAL_PRODUCT = "DIGITAL_PRODUCT"


class ProductStatus(StrEnum):
    """Supported product lifecycle statuses."""

    DRAFT = "DRAFT"
    ACTIVE = "ACTIVE"
    INACTIVE = "INACTIVE"
    ARCHIVED = "ARCHIVED"


class ProductSchema(BaseModel):
    """Strict API schema configuration for product module."""

    model_config = ConfigDict(extra="forbid", from_attributes=True)


class ProductCategoryCreate(ProductSchema):
    """Create payload for one hierarchical product category node."""

    code: str = Field(min_length=2, max_length=50, pattern=r"^[A-Z0-9_-]+$")
    name: str = Field(min_length=1, max_length=200)
    parent_id: UUID | None = None
    is_active: bool = True

    @field_validator("code", mode="before")
    @classmethod
    def normalize_code(cls, value: str) -> str:
        """Normalize category code for deterministic matching."""
        return value.strip().upper()


class ProductCategoryUpdate(ProductCategoryCreate):
    """Replace payload for one category node."""


class ProductCategoryResponse(ProductSchema):
    """Response contract for one category node."""

    id: UUID
    firm_id: UUID
    code: str
    name: str
    parent_id: UUID | None
    level: int
    path: str
    is_active: bool
    created_at: datetime
    updated_at: datetime


class ProductAttributeInput(ProductSchema):
    """One dynamic attribute value submitted for a product."""

    attribute_definition_id: UUID
    value: str | int | float | bool | date


class ProductMediaInput(ProductSchema):
    """One product media or document reference."""

    media_kind: str = Field(min_length=2, max_length=30)
    file_name: str = Field(min_length=1, max_length=260)
    mime_type: str | None = Field(default=None, max_length=120)
    storage_path: str = Field(min_length=1, max_length=1024)
    is_primary: bool = False
    file_size_bytes: int | None = Field(default=None, ge=0)

    @field_validator("media_kind", mode="before")
    @classmethod
    def normalize_media_kind(cls, value: str) -> str:
        """Normalize media kind token to uppercase."""
        return value.strip().upper()


class ProductWrite(ProductSchema):
    """Shared writable payload for create and update."""

    code: str = Field(min_length=2, max_length=50, pattern=r"^[A-Z0-9_-]+$")
    barcode: str | None = Field(default=None, max_length=120)
    qr_code: str | None = Field(default=None, max_length=300)
    name: str = Field(min_length=1, max_length=200)
    short_name: str | None = Field(default=None, max_length=100)
    description: str | None = None
    product_type: ProductType
    category_id: UUID | None = None
    sub_category_id: UUID | None = None
    unit: str | None = Field(default=None, max_length=20)
    brand: str | None = Field(default=None, max_length=120)
    model: str | None = Field(default=None, max_length=120)
    hsn_sac: str | None = Field(default=None, max_length=20)
    tax_profile_group_code: str | None = Field(
        default=None, max_length=50, pattern=r"^[A-Z0-9_-]+$"
    )
    base_uom_id: UUID | None = None
    inventory_uom_id: UUID | None = None
    purchase_uom_id: UUID | None = None
    sales_uom_id: UUID | None = None
    default_receiving_uom_id: UUID | None = None
    default_dispatch_uom_id: UUID | None = None
    minimum_sales_uom_id: UUID | None = None
    weight: Decimal | None = Field(default=None, ge=0, max_digits=18)
    volume: Decimal | None = Field(default=None, ge=0, max_digits=18)
    length: Decimal | None = Field(default=None, ge=0, max_digits=18)
    width: Decimal | None = Field(default=None, ge=0, max_digits=18)
    height: Decimal | None = Field(default=None, ge=0, max_digits=18)
    allow_fraction: bool = False
    allow_decimal: bool = True
    purchase_price: Decimal | None = Field(default=None, ge=0, max_digits=18)
    selling_price: Decimal | None = Field(default=None, ge=0, max_digits=18)
    mrp: Decimal | None = Field(default=None, ge=0, max_digits=18)
    status: ProductStatus = ProductStatus.ACTIVE
    remarks: str | None = None
    track_batch: bool = False
    track_lot: bool = False
    track_serial: bool = False
    track_expiry: bool = False
    track_manufacturing_date: bool = False
    track_warranty: bool = False
    allow_negative_stock: bool = False
    require_batch_on_receipt: bool = False
    require_batch_on_issue: bool = False
    require_serial_on_receipt: bool = False
    require_serial_on_issue: bool = False
    attributes: list[ProductAttributeInput] = Field(
        default_factory=list, max_length=300
    )
    media: list[ProductMediaInput] = Field(default_factory=list, max_length=200)

    @field_validator("code", "unit", "hsn_sac", mode="before")
    @classmethod
    def normalize_upper(cls, value: str | None) -> str | None:
        """Normalize selected identifier-like fields to uppercase."""
        if value is None:
            return None
        normalized = value.strip().upper()
        return normalized or None

    @model_validator(mode="after")
    def validate_pricing(self) -> "ProductWrite":
        """Ensure selling price never exceeds MRP."""
        if (
            self.mrp is not None
            and self.selling_price is not None
            and self.mrp < self.selling_price
        ):
            raise ValueError("MRP must be greater than or equal to selling price.")
        return self


class ProductCreate(ProductWrite):
    """Create a product core record and all dynamic extensions."""


class ProductUpdate(ProductWrite):
    """Replace all editable product fields and dynamic values."""


class ProductImportRequest(ProductSchema):
    """Import request for JSON-based product batches."""

    records: list[ProductCreate] = Field(min_length=1, max_length=3000)


class ProductAttributeResponse(ProductSchema):
    """Persisted attribute value response."""

    id: UUID
    attribute_definition_id: UUID
    value_text: str | None
    value_number: Decimal | None
    value_date: date | None
    value_boolean: bool | None
    created_at: datetime
    updated_at: datetime


class ProductMediaResponse(ProductSchema):
    """Persisted media/document response."""

    id: UUID
    media_kind: str
    file_name: str
    mime_type: str | None
    storage_path: str
    is_primary: bool
    file_size_bytes: int | None
    created_at: datetime
    updated_at: datetime


class ProductResponse(ProductSchema):
    """Full product response envelope payload."""

    id: UUID
    firm_id: UUID
    code: str
    barcode: str | None
    qr_code: str | None
    name: str
    short_name: str | None
    description: str | None
    product_type: ProductType
    category_id: UUID | None
    sub_category_id: UUID | None
    unit: str | None
    brand: str | None
    model: str | None
    hsn_sac: str | None
    tax_profile_group_code: str | None
    base_uom_id: UUID | None
    inventory_uom_id: UUID | None
    purchase_uom_id: UUID | None
    sales_uom_id: UUID | None
    default_receiving_uom_id: UUID | None
    default_dispatch_uom_id: UUID | None
    minimum_sales_uom_id: UUID | None
    weight: Decimal | None
    volume: Decimal | None
    length: Decimal | None
    width: Decimal | None
    height: Decimal | None
    allow_fraction: bool
    allow_decimal: bool
    purchase_price: Decimal | None
    selling_price: Decimal | None
    mrp: Decimal | None
    status: ProductStatus
    remarks: str | None
    track_batch: bool
    track_lot: bool
    track_serial: bool
    track_expiry: bool
    track_manufacturing_date: bool
    track_warranty: bool
    allow_negative_stock: bool
    require_batch_on_receipt: bool
    require_batch_on_issue: bool
    require_serial_on_receipt: bool
    require_serial_on_issue: bool
    is_deleted: bool
    deleted_at: datetime | None
    created_by: UUID | None
    created_at: datetime
    updated_by: UUID | None
    updated_at: datetime
    # Resolved from the shared attribute store by the router, not the ORM row.
    attributes: list[ProductAttributeResponse] = Field(default_factory=list)
    media: list[ProductMediaResponse]


class ProductSummary(ProductSchema):
    """Aggregate product list summary."""

    total: int
    active: int
    inactive: int
    draft: int
    archived: int
    deleted: int


class ProductListFilters(ProductSchema):
    """Validated filters for listing products."""

    status: ProductStatus | None = None
    product_type: ProductType | None = None
    category_id: UUID | None = None
    sub_category_id: UUID | None = None
    tax_profile_group_code: str | None = None
    brand: str | None = Field(default=None, max_length=120)
    hsn_sac: str | None = Field(default=None, max_length=20)
    include_deleted: bool = False
    attribute_query: str | None = Field(default=None, max_length=200)


class ProductCategoryFilter(ProductSchema):
    """Category tree list filter."""

    parent_id: UUID | None = None
    include_inactive: bool = False


class ProductFeatureState(ProductSchema):
    """Resolved feature flag state for current firm/profile."""

    code: str
    enabled: bool


class ProductTaxProfileOption(ProductSchema):
    """Tax profile option metadata for product tax selection."""

    id: UUID
    code: str
    group_code: str | None
    label: str
    tax_system_id: UUID


class ProductMetadataResponse(ProductSchema):
    """Dynamic metadata used by desktop product forms."""

    profile_code: str
    features: list[ProductFeatureState]
    categories: list[ProductCategoryResponse]
    tax_profiles: list[ProductTaxProfileOption]
    required_attribute_definition_ids: list[UUID]
    optional_attribute_definition_ids: list[UUID]


class BulkProductRequest(ProductSchema):
    """Bulk operation target identifiers."""

    ids: list[UUID] = Field(min_length=1, max_length=2000)
